### PR TITLE
Python coverage

### PR DESCRIPTION
**Pull request by hsorby**
Adding coverage testing for the Python bindings.  Python bindings will now need to have 100% coverage for all API calls.

Addresses #253.

### CODE REVIEW

**Comment by kerimoyle** [comment]
I thought we were only working towards coverage at the moment, that Python API coverage wouldn't prevent merging just yet? Or has that changed?

**Comment by hsorby** [comment]
I have improved Python API coverage to 100% so there is no need for it to prevent merging.

**Comment by hsorby** [comment]
Sorry team, I have added a check in the Buildbot CI for 100% Python API coverage.  This means that this PR will need to be dealt with fairly soonish if we want to see some green ticks on other PRs.  Again apologies for the inconvenience that this causes.

**Comment by nickerso** [comment]
Does this one need to be re-triggered to perform that Python coverage test? or is it hidden inside one of the other checks?

File: CMakeLists.txt [hunk 1]
```diff
@@ -100,14 +100,28 @@ unset(MEMCHECK CACHE)
...
 set(_PARAM_ANNOTATION "Build Python wrappers.")
-if(BINDINGS_AVAILABLE)
+if(BINDINGS_AVAILABLE AND Python_Development_FOUND)
   set(LIBCELLML_BINDINGS_PYTHON ON CACHE BOOL ${_PARAM_ANNOTATION})
 endif()
-if(DEFINED BINDINGS_PYTHON AND BINDINGS_AVAILABLE)
+if(DEFINED BINDINGS_PYTHON AND BINDINGS_AVAILABLE AND Python_Development_FOUND)
   set(LIBCELLML_BINDINGS_PYTHON "${BINDINGS_PYTHON}" CACHE BOOL ${_PARAM_ANNOTATION} FORCE)
+elseif(BINDINGS_PYTHON)
+  message(WARNING "Python bindings requested but development files for Python or swig was not found!")
```
Comment (agarny) [inline CMakeLists.txt]: Should be `SWIG`, not `swig`.

File: CMakeLists.txt [hunk 2]
```diff
@@ -100,14 +100,28 @@ unset(MEMCHECK CACHE)
...
 
+# PYTHON_COVERAGE ==> LIBCELLML_PYTHON_COVERAGE
+set(_PARAM_ANNOTATION "Enable Python coverage testing.")
+if(PYTHON_COVERAGE_TESTING_AVAILABLE)
+  set(LIBCELLML_PYTHON_COVERAGE ON CACHE BOOL ${_PARAM_ANNOTATION})
+endif()
+if(DEFINED PYTHON_COVERAGE AND PYTHON_COVERAGE_TESTING_AVAILABLE)
+  set(LIBCELLML_PYTHON_COVERAGE "${PYTHON_COVERAGE}" CACHE BOOL ${_PARAM_ANNOTATION} FORCE)
+elseif(PYTHON_COVERAGE)
+  message(WARNING "Python coverage testing requested but the requirements were not found!")
```
Comment (agarny) [inline CMakeLists.txt]: Such a warning wouldn't help me unless I was to check this file. Might be worth mentioning what those requirements are? (Same elsewhere maybe?)

File: src/api/libcellml/component.h [hunk 3]
```diff
@@ -416,21 +416,20 @@ class LIBCELLML_EXPORT Component: public ComponentEntity, public ImportedEntity
...
      * @param importSource The @c ImportSourcePtr to add to this @c Component.
      */
-    void setImportSource(ImportSourcePtr &importSource);
-
-private:
-    Component(); /**< Constructor @private*/
-    explicit Component(const std::string &name); /**< Constructor named @private */
-
-    bool doAddComponent(const ComponentPtr &component) override; /**< Virtual method for implementing addComponent, @private */
+    void doSetImportSource(const ImportSourcePtr &importSource) override;
```
Comment (agarny) [inline src/api/libcellml/component.h:432]: Just as a reminder, `doAddComponent()` and `doSetImportSource()` will have to be removed from the API at some point. They really have nothing to do here.

File: src/api/libcellml/importedentity.h [hunk 4]
```diff
@@ -98,6 +98,16 @@ class LIBCELLML_EXPORT ImportedEntity
...
 
+    /**
+     * @brief Virtual set import source method to be implemented by derived classes.
+     *
+     * Virtual setImportSource method to allow the units and component classes to
+     * implement their own versions.
+     *
+     * @param importSource The import source to set.
+     */
+    virtual void doSetImportSource(const ImportSourcePtr &importSource);
```
Comment (agarny) [inline src/api/libcellml/importedentity.h:109]: Just as a reminder, `doSetImportSource()` will have to be removed from the API at some point. It really has nothing to do here.

File: src/bindings/interface/componententity.i [hunk 5]
```diff
@@ -79,10 +79,13 @@ the structure."
 
 
 #if defined(SWIGPYTHON)
+    %ignore libcellml::ComponentEntity::replaceComponent(size_t index, const ComponentPtr &component);
+
     // Allow any type of input to be converted to bool
     %typemap(typecheck,precedence=SWIG_TYPECHECK_BOOL) bool { $1 = 1; }
     %typemap(in) bool { $1 = PyObject_IsTrue($input) == 1; }
 
+
```
Comment (agarny) [inline src/bindings/interface/componententity.i]: Why this empty line?

File: src/api/libcellml/units.h [hunk 6]
```diff
@@ -492,19 +492,20 @@ class LIBCELLML_EXPORT Units: public NamedEntity, public ImportedEntity
...
      *
-     * @param importSource The @c ImportSourcePtr to add to this @c Units item.
+     * @param importSource The @c ImportSourcePtr to add to this @ref Units.
      */
-    void setImportSource(ImportSourcePtr &importSource);
-
-private:
-    Units(); /**< Constructor */
-    explicit Units(const std::string &name); /**< Constructor with std::string parameter*/
+    void doSetImportSource(const ImportSourcePtr &importSource) override;
```
Comment (agarny) [inline src/api/libcellml/units.h:508]: Just as a reminder, `doSetImportSource()` will have to be removed from the API at some point. It really has nothing to do here.

**Review by agarny — COMMENTED** [review]
Just some minor comments. Looks good otherwise.

**Review by nickerso — DISMISSED** [review]
having now realised that the Python coverage is being tested, its just hidden under the GCov coverage test, I'm happy to approve this one.

**Comment by nickerso** [comment]
> Sorry team, I have added a check in the Buildbot CI for 100% Python API coverage. This means that this PR will need to be dealt with fairly soonish if we want to see some green ticks on other PRs. Again apologies for the inconvenience that this causes.

Can we run the Python coverage testing on Windows so that we would then have some form of coverage across all three platforms?

**Comment by hsorby** [comment]
> > Sorry team, I have added a check in the Buildbot CI for 100% Python API coverage. This means that this PR will need to be dealt with fairly soonish if we want to see some green ticks on other PRs. Again apologies for the inconvenience that this causes.
> 
> Can we run the Python coverage testing on Windows so that we would then have some form of coverage across all three platforms?

We can but I put it onto the Ubuntu test machine as that one finishes the earliest.  The machine that the Python coverage testing is done on shouldn't (I say shouldn't) make a difference.

**Review by nickerso — APPROVED** [review]

**Review by agarny — APPROVED** [review]

**Pull request merged**